Support non-UTC time zones in Timeout assertions

philippdolder opened the pull request:
This PR is intended as a way to demonstrate/discuss what I think is a minor bug in the "old" testing framework. Assertions on non-UTC date times don't work

Change Timeout expectations from DateTime to DateTimeOffset in order to allow assertions on non-UTC time zones.

andreasohlund commented:
What's the actual bug? (should one of the new tests fail?)

DavidBoike commented:
@philippdolder Sorry I didn't see it before on Gitter, but I would only ever use UTC times and UtcNow for timeouts. If anything, I think the bug is that we don't throw an exception when given a Local DateTime instance. (A timeout is a request for something to happen at a specific point in time, regardless of location on Earth, so a Utc DateTime is most appropriate.)

In any case, we wouldn't be able to change the public API for all of that to DateTimeOffset (a breaking API change) except in a major version.

philippdolder commented:
@andreasohlund Sorry, the new tests fail without my change in the Testing framework.

philippdolder commented:
@DavidBoike I agree that one should only ever use UTC dates, not only for timeouts, but for everything that's stored somewhere. Unfortunately our codebase uses DateTime.Now all over the place, which isn't a simple change.

As the real implementation accepts a Local DateTime the testing framework should also support it. Though of course one could discuss if the real implementation should accept local DateTime at all. That would certainly be a breaking change.

Regarding the breaking change nature of my proposal: all the existing code should work fine as DateTimeOffset has an implicit conversion from DateTime taking into account the TimeZone Offset. Of course if somebody is explicitly casting between the two. But I'm not extremely experienced with this kind of public API changes. I know you're the experts, just wanted to explain my thoughts.

timbussmann commented:
I think the proposed changes are the correct approach to fix this as the `SendOptions` API is designed to work with `DateTimeOffset`. Therefore it's incorrect to specify the checks on the testing framework only as `DateTime`. Sadly, this change is a binary breaking change (you have to recompile the binaries in order for the implicit casting to take effect). I think we need to discuss the steps in regard to the upcoming v7 RC @Particular/nservicebus-maintainers.

there is a workaround to get the test passing by slightly adjusting the expectation to:
`.ExpectDefer<TestMessage>((m, t) => t.ToLocalTime() == datetime)` as this converts the `DateTime` parameter into the correct `DateTime.Kind` matching the uses local `DateTime.Now`. casting a `DateTimeOffset` to `DateTime` seems to always set the Kind to `unspecified` for some reason.

philippdolder commented:
Thank you @timbussmann for the insight. I didn't think about the need to recompile.
I have a working solution for my tests and try to move towards the new testing framework.

I'm fine if you like to close the PR if you don't need it anymore for the v7 discussion mentioned by @timbussmann

timbussmann commented:
@philippdolder I'd prefer to pull your changes into the v7 rc but I need to discuss this with the maintainers first (as the next window for this change as proposed right now would be v8), so no need to close it yet :) Not sure how we're going to address this for v6 though tbh.

timbussmann commented:
Hey @philippdolder 
We would love to pull in your PR, but for the reasons of breaking changes mentioned before, this PR needs to go into the v7 release of NServiceBus.Testing. Would it possible for you to rebase your branch on top of the `release-7.0` branch and adjust the target branch to `release-7.0` as well?
As for v6, we would have to add new overloads instead of replacing the existing expectations. I'm not really sure this would solve the issue though, as your tests would still fallback to the current API working with `DateTime` directly and therefore continue to fail. As there is a workaround available, we think it's not very critical to backport this, what are your thoughts on that?

have you already signed our [contributors license agreement](https://particular.net/contributors-license-agreement-consent)?

philippdolder commented:
@timbussmann I've rebased on release-7.0 and adjusted the target branch accordingly.
I also just signed the contributors license agreement.
And I removed the WIP prefix from the PR title.

Re fixing it for v6. I agree with you. It won't help to have overloads to mitigate the issue.

timbussmann commented:
merging this then, thanks a lot for this solid PR @philippdolder ❤️ Those changes will be released as part of 7.0.0-Beta5 of NServiceBus.Testing (pushing this very soon to our myget feed).
